### ISSUE TITLE
Clarify purpose of "aggregate-results.sh"

### ISSUE THREAD

**Issue by sumpfralle** [felipec/sharness#78]
Hello,

I am about to package *sharness* for Debian.
But I am not sure, if (and how) I should add "aggregate-results.sh" to the package.
At the moment, I see the following options:
* A) do not include the file in the package
* B) `/usr/share/sharness/aggregate-results.sh`: this would be a bit unusual for an executable script
* C) `/usr/share/doc/sharness/examples/aggregate-results.sh`: this would be the proper location for an example-like snippet
* D) `/usr/bin/sharness-aggregate-results`: I would like to avoid this deviation from upstream (you use a different filename) and I am afraid, that the existence of an executable within the package would confuse new users (and prevent them from reading the docs first). Is this script important enough to be part of the system PATH?

Cheers,
Lars

**Comment by chriscool** [comment]
Would `/usr/share/sharness/bin/aggregate-results.sh` be possible?

I wonder if you will also provide the [Makefile](https://github.com/chriscool/sharness/blob/master/test/Makefile) or if you expect people to write their own Makefile or build script to launch aggregate-results.sh?

In the README.md we suggest:

> If you like to add Sharness to the sources of a project you want to use it for, simply copy the files sharness.sh, aggregate-results.sh, and test/Makefile to a folder named test inside that project.

Do you expect people to copy those files from the location where the Debian package install it? I guess not but maybe that will be what some people will want to do.

**Comment by sumpfralle** [comment]
> Would /usr/share/sharness/bin/aggregate-results.sh be possible?

I think, this would be misleading, since my understanding of the FHS (and the [Debian policy](https://www.debian.org/doc/debian-policy/#shared-library-support-files)) forbids executable files below `/usr/share`. Thus I would currently prefer `/usr/share/doc/sharness/contrib/aggregate-results.sh` as a location indicating the optional nature of that file and its source (upstream).

> I wonder if you will also provide the Makefile or if you expect people to write their own Makefile or build script to launch aggregate-results.sh?

My main focus of the first packaging version aimed at the usage pattern, that I used until now: running tests as part of Debian's `autopkgtests` as defined by [DEP8](https://dep-team.pages.debian.net/deps/dep8/). I described this usage pattern in the [package documentation](https://salsa.debian.org/debian/sharness/blob/master/debian/sharness.README.Debian).

But of course, we can expand starting from there.

Thus let us explore, which kind of other usage pattern we could envision.

Regarding the integration of tests within software that I maintain, I could imagine to add the following lines to my (existing) Makefile:
```make
TESTS = $(sort $(wildcard $(tests/*.t))

test:
        prove $(TESTS)
```

Would the above example lack notable features?
If yes: do you have a suggestion, how we could simplify these for the users of the Debian package?

Regarding your example Makefile (`tests/Makefile`): I could imagine, that we want to ship it as `/usr/include/sharness/sharness_test.mk`. This would allow users to include it in their Makefiles. Do you think, that this would require changes of the variables or target names in order to avoid collisions with the namespace of the parent (including) Makefile?

Regarding `aggregate-results.sh`: if you think, it is a useful feature, that a significant number of `sharness` users will want to run (as a script), then we should move it to the PATH (e.g. `/usr/bin/sharness-aggregate-results`). It is a bit of a pity, that this name would deviate from yours (upstream), but the name `aggregate-results.sh` would [collide with the Debian policy due to its `.sh` extension](https://www.debian.org/doc/debian-policy/#scripts) and it would be a confusing name for a system-wide executable, I think.

Please share your thoughts, so we can craft a packaging, that you are happy with ...

**Comment by chriscool** [comment]
Ok for `/usr/share/doc/sharness/contrib/aggregate-results.sh`.

> Would the above example lack notable features?

The output summary might be a bit better if after `prove $(TESTS)` something like the following (from `tests/Makefile`) was used:

```
aggregate-results:
        for f in test-results/*.counts; do \
                echo "$$f"; \
        done | '$(SHELL_PATH_SQ)' '$(AGGREGATE_SCRIPT)'
```

and unfortunately it's not so easy to find that the aggregate script should be used like that.

> If yes: do you have a suggestion, how we could simplify these for the users of the Debian package?

I think that your suggestion to ship `tests/Makefile` as `/usr/include/sharness/sharness_test.mk` is probably the best solution. It documents how the aggregate script should and could be used and makes it easy to use it and Sharness in general. For example it makes it easy to run the tests without `prove`.

> Do you think, that this would require changes of the variables or target names in order to avoid collisions with the namespace of the parent (including) Makefile?

Yeah, probably, and I am ok with doing that and also maybe extracting `sharness_test.mk` from `tests/Makefile`.

**Comment by felipec** [comment]
For what it's worth I've been using sharness for many years and never have I used `aggregate-results.sh`, as I don't see much value in it.

Perhaps seeing that there's a new use case that is now fixed would be useful, but that can be done in a much simpler script.

Anyway, nowadays `make install` installs `aggregate-results.sh` into `$(prefix)/share/sharness/lib-sharness`, so this issue is mostly fixed, no? I think it's the wrong location since it's not really a library, but a tool (which isn't necessary), so maybe `/share/sharness/tools`, or `helpers`.

Is there anything still to do here?

**Comment by chriscool** [comment]
I agree that the fact that `make install` installs `aggregate-results.sh` is fixing part of the issue. Now I am not sure if `make install` also installs a Makefile that makes `aggregate-results.sh` easy to use. That could be an improvement that I think could fix the rest of this issue if it's not the case yet.

**Comment by felipec** [comment]
There's a few things I can immediately think of.

First, the Makefile doesn't need a for loop just to list files, `ls -1` does the trick:

    ls -1 test-results/*.counts | '$(SHELL_PATH_SQ)' '$(AGGREGATE_SCRIPT)'

Second, the reason the list of files is passed to the script is that IRIX 6.5 had an argument limitation of 20480, but we can just pass the filter as a string `'test-results/*.counts'`.

    for file in $1; do

And then, most projects aren't even going to specify anything else, so:

    for file in ${1-test-results/*.counts}; do

That would make the script much easier to use.

**Comment by chriscool** [comment]
Yeah, maybe documenting that at the top of `aggregate-results.sh` would be enough then.

**Comment by felipec** [comment]
There's not much to document if the script just works. See: 743efc3.

**Comment by felipec** [comment]
FWIW, in git upstream they like my proposed changes: [test: simplify counts aggregation](https://lore.kernel.org/git/CAMP44s2NJefUOzpyq=qbWSi+X-GnVTXaEgT8dM4zMPSzm29yag@mail.gmail.com/T/#m917a8fbc2ee1553c40e5e6d345ee6ab7c7591886).

**Comment by felipec** [comment]
Upstream [seemed to like](https://lore.kernel.org/git/20230310215113.M656422@dcvr/) my proposal, so I went ahead and pushed something similar: 5e38ec8.

Now it's much easier to explain the purpose of the script, but I have some stuff pending to explain everything in a new Wiki page.

**Comment by felipec** [comment]
Dealing with `aggregate-results.sh` resulted in many changes.

1. I moved `aggregate-results.sh` to a `tools/` directory
2. I created an `example/` directory to show basic usage of Sharness
3. I removed `aggregate-results.sh` from the `example/Makefile` to make it simpler (it's not necessary)
4. I explained in `README.md` that `aggregate-results.sh` is not necessary, but a tool to be called in the project's `Makefile` (if the project wishes to do so)

It should be clear now that this is a tool, and it's **optional**. Perhaps if enough people push, we could create a more complex example that does use this script, but for now I think I think it's clearly explained.

BTW. In my Arch Linux package it's placed in `/usr/share/sharness/tools/` (as that's the default with `prefix=/usr`).
